Feat halo

dhinf opened the pull request:
get HaloMatrixWrapper out of experimental
- adapted tests and examples
- removed old Halo.h
- removed GlobStencilIter + tests for GlobStencilIter (Tobias please have look at it)

dhinf committed:
first version of new halo; still buggy

dhinf committed:
first version of new halo; still buggy

dhinf committed:
first version of new halo; still buggy

dhinf committed:
first version of new halo; still buggy

dhinf committed:
MatrixHaloWrapper now with StencilSpec init

dhinf committed:
MatrixHaloWrapper now with StencilSpec init

dhinf committed:
MatrixHaloWrapper now with StencilSpec init

dhinf committed:
MatrixHaloWrapper now with StencilSpec init

dhinf committed:
added stencilSpec to HaloMatrixIterators

dhinf committed:
added stencilSpec to HaloMatrixIterators

dhinf committed:
added stencilSpec to HaloMatrixIterators

dhinf committed:
added stencilSpec to HaloMatrixIterators

dhinf committed:
type_traits for pow

dhinf committed:
type_traits for pow

dhinf committed:
type_traits for pow

dhinf committed:
type_traits for pow

dhinf committed:
fixed region index calculation in valueAt and removed code duplication

dhinf committed:
fixed region index calculation in valueAt and removed code duplication

dhinf committed:
fixed region index calculation in valueAt and removed code duplication

dhinf committed:
fixed region index calculation in valueAt and removed code duplication

dhinf committed:
added fixed halos, some bugfixes

dhinf committed:
added fixed halos, some bugfixes

dhinf committed:
added fixed halos, some bugfixes

dhinf committed:
added fixed halos, some bugfixes

dhinf committed:
Merge remote-tracking branch 'upstream/development' into feat-halo

dhinf committed:
Merge branch 'development' of https://github.com/dash-project/dash into feat-halo

dhinf committed:
adapted Halo stuff to new memory structure

dhinf committed:
adapted Halo stuff to new memory structure

dhinf committed:
adapted Halo stuff to new memory structure

dhinf committed:
adapted Halo stuff to new memory structure

dhinf committed:
Merge branch 'development' of https://github.com/dash-project/dash into feat-halo

dhinf committed:
added RegionCoords

dhinf committed:
added RegionCoords

dhinf committed:
added RegionCoords

dhinf committed:
added RegionCoords

dhinf committed:
bugfixes boundary elements and stancil offsets

dhinf committed:
bugfixes boundary elements and stancil offsets

dhinf committed:
bugfixes boundary elements and stancil offsets

dhinf committed:
bugfixes boundary elements and stancil offsets

dhinf committed:
bug fixes

dhinf committed:
bug fixes

dhinf committed:
bug fixes

dhinf committed:
bug fixes

dhinf committed:
renamed HaloMatrix to HaloMatrixWrapper

dhinf committed:
renamed HaloMatrix to HaloMatrixWrapper

dhinf committed:
renamed HaloMatrix to HaloMatrixWrapper

dhinf committed:
renamed HaloMatrix to HaloMatrixWrapper

dhinf committed:
fixed region spec calculation for stencils

dhinf committed:
fixed region spec calculation for stencils

dhinf committed:
fixed region spec calculation for stencils

dhinf committed:
fixed region spec calculation for stencils

dhinf committed:
added tests for halos

dhinf committed:
added tests for halos

dhinf committed:
added tests for halos

dhinf committed:
Merge branch 'development' of https://github.com/dash-project/dash into feat-halo

dhinf committed:
Merge branch 'feat-halo' of https://github.com/dash-project/dash into feat-halo

dhinf committed:
changed  include to HaloMatrixWrapper

dhinf committed:
changed  include to HaloMatrixWrapper

dhinf committed:
added Stencil to HaloSpec calculation tests

dhinf committed:
added Stencil to HaloSpec calculation tests

dhinf committed:
minor changes in 2D test and cout removed in Wrapper

dhinf committed:
minor changes in 2D test and cout removed in Wrapper

dhinf committed:
added dart_get_handled for strides and indexes

dhinf committed:
added dart_get_handled for strides and indexes

dhinf committed:
adapted HaloWrapper to new dart_get methodes and bug fixes especially for Col_Major

dhinf committed:
adapted HaloWrapper to new dart_get methodes and bug fixes especially for Col_Major

dhinf committed:
new tests for dart_get strided and indexed

dhinf committed:
new tests for dart_get strided and indexed

dhinf committed:
added Col_Major test for Halos

dhinf committed:
added Col_Major test for Halos

dhinf committed:
bugfixes

dhinf committed:
bugfixes

dhinf committed:
Merge branch 'feat-halo' of https://github.com/dash-project/dash into feat-halo

dhinf committed:
fixed minor bugs, added method to access halo elements locally via coordinates

dhinf committed:
Merge remote-tracking branch 'upstream/development' into feat-halo

dhinf committed:
adapted indexed/strided functions and HaloMatrixWrapper to changes made in dart

dhinf committed:
fixed num requests in srided/indexed handled get and wait_local for halo_update

dhinf committed:
moved halo stuff out of experimental

moved halo stuff out of experimental

moved halo stuff out of experimental

dhinf committed:
moved halo stuff out of experimental

dhinf committed:
Merge remote-tracking branch 'upstream/development' into feat-halo

dhinf committed:
Merge branch 'feat-halo' of https://github.com/dash-project/dash into feat-halo

devreal commented on dart-impl/mpi/src/dart_communication.c:
```diff
@@ -550,6 +550,317 @@ dart_ret_t dart_compare_and_swap(
+  DART_LOG_TRACE("dart_get_strided_handle > handle(%p) dest:%d win:%"PRIu64,
+                 (void*) handle, handle->dest, (unsigned long)win);
+
+  MPI_Type_free(&mpi_strided_type);
```
Please make sure the stride is free'd in all cases (including errors).

devreal commented on dart-if/include/dash/dart/if/dart_communication.h:
```diff
@@ -473,46 +516,46 @@ typedef struct dart_handle_struct * dart_handle_t;
- * \threadsafe
- * \ingroup DartCommunication
- */
+void            * dest,
```
Please undo the unindent (probably done by your editor)

devreal reviewed (changes requested):
Only looked at the DART part for now, please see my two cents inline. I will look at the rest later

dhinf committed:
better error handling for strided/indexed, typo fixes

devreal commented on dart-if/include/dash/dart/if/dart_communication.h:
```diff
@@ -473,11 +516,11 @@ typedef struct dart_handle_struct * dart_handle_t;
-  size_t            nelem,
-  dart_datatype_t   dtype,
-  dart_handle_t   * handle) DART_NOTHROW;
+void            * dest,
```
Some more whitespaces accidentally removed :)

devreal commented on dash/include/dash/halo/Halo.h:
```diff
@@ -0,0 +1,1062 @@
+ */
+enum class Cycle : uint8_t {
+  /// No global boundary Halos
+  NONE,
```
Are we sure there is no name conflict on `dash::NONE`? Can we move this to `dash::Stencil::NONE`?

dhinf replied:
it's an enum class -> you can call it with dash::Cycle::NONE only

devreal replied:
Ahh, didn't see that (and wasn't aware of enum classes to be honest ^^)

devreal commented on dash/include/dash/halo/Halo.h:
```diff
@@ -0,0 +1,1062 @@
+    CoordsT coords{};
+    region_index_t  index_tmp = static_cast<long>(index);
+    for (auto i = (NumDimensions - 1); i >= 1; --i) {
+      auto res  = std::div(index_tmp, 3);
```
Where does the constant `3` come from?

dhinf replied:
it's i fixed constant -> i have dimension^3 regions(including the center)
e.g. 1D 3 regions: north, center, south
       2D 9 regions: north west, north, north east, west, center, east,south west, south, south east
etc.

devreal replied:
Mhh, but in 1D 1^3 is 1 ;)

In any case, I would suggest using a `constexpr` variable to give the value `3` a meaningful name.

dhinf replied:
meant 3^dimension. sorry

dhinf replied:
constexpr THREE = 3 :D

dhinf replied:
you are right. i will find a meaningful name

devreal commented on dash/include/dash/halo/Halo.h:
```diff
@@ -0,0 +1,1062 @@
+
+  template<typename... ARGS>
+  HaloSpec(const ARGS&... args) {
+    std::vector<HaloRegionSpecT> tmp{args...};
```
Better use `std::array` here

devreal commented on dash/include/dash/halo/Halo.h:
```diff
@@ -0,0 +1,1062 @@
+private:
+  void setRegionSpec(const StencilT& stencil) {
+        auto index = HaloRegionSpecT::index(stencil);
+        auto max = stencil.max();
```
Indentation!

devreal commented on dash/include/dash/halo/Halo.h:
```diff
@@ -0,0 +1,1062 @@
+   * regions.
+   */
+  RegionIter(GlobMemT& globmem, const PatternT& pattern, const ViewSpecT& _region_view,
+             index_type pos, index_type size)
```
`index_type size` -> `size_type size`

devreal commented on dash/include/dash/halo/Halo.h:
```diff
@@ -0,0 +1,1062 @@
+  /**
+   * Copy constructor.
+   */
+  RegionIter(const SelfT& other) = default;
```
What about move assignment/construction?

devreal commented on dash/include/dash/halo/Halo.h:
```diff
@@ -0,0 +1,1062 @@
+  RegionIter(GlobMemT& globmem, const PatternT& pattern, const ViewSpecT& _region_view,
+             index_type pos, index_type size)
+      : _globmem(globmem), _pattern(pattern), _region_view(_region_view), _idx(pos), _size(size),
+        _max_idx(_size - 1), _myid(dash::myid()), _lbegin(globmem.lbegin()) {}
```
The iterator should be team-aware, use `pattern.team.myid()` instead of `dash::myid()`

devreal commented on dash/include/dash/halo/Halo.h:
```diff
@@ -0,0 +1,1062 @@
+    return res;
+  }
+
+  index_type operator+(const SelfT& other) const { return _idx + other._idx; }
```
Formatting! 

Plus: why does this `operator+` return `index_type`?

Same holds true for the below operator.

devreal commented on dash/include/dash/halo/Halo.h:
```diff
@@ -0,0 +1,1062 @@
+  /// Maximum iterator position in the block border's iteration space.
+  index_type _max_idx{0};
+  /// Unit id of the active unit
+  dart_unit_t _myid;
```
Use `dash::team_unit_t` here

devreal commented on dash/include/dash/halo/Halo.h:
```diff
@@ -0,0 +1,1062 @@
+  Region(const HaloRegionSpecT& region_spec, const ViewSpecT& region, GlobMemT& globmem,
+         const PatternT& pattern, const BorderT& border)
+      : _region_spec(region_spec), _region(region), _border(border),
+        _border_region(std::any_of(border.begin(), border.end(), [](bool border_dim){ return border_dim == true;})),
```
line length >80 chars?

dhinf replied:
120

devreal replied:
Please stick to 80 :)

dhinf replied:
but we don't have IBM punched cards anymore and instead nice wide screens :)

devreal commented on dash/include/dash/halo/Halo.h:
```diff
@@ -0,0 +1,1062 @@
+template <typename ElementT, typename PatternT>
+class HaloBlock {
+private:
+  static const dim_t NumDimensions = PatternT::ndim();
```
Can this be `constexpr`?

devreal commented on dash/include/dash/halo/Halo.h:
```diff
@@ -0,0 +1,1062 @@
+
+  size_type halo_size() const {
+    return std::accumulate(_halo_regions.begin(), _halo_regions.end(), 0,
+                           [](size_type sum, region_t region) { return sum + region.size(); });
```
Pass `region` by reference?

devreal commented on dash/include/dash/halo/HaloMatrixWrapper.h:
```diff
@@ -0,0 +1,307 @@
+
+  void updateHaloIntern(Data& data, bool async) {
+    auto rel_dim = data.region.regionSpec().relevantDim() - 1;
+    if (data.region.borderRegion() && data.region.borderDim(rel_dim) && _cycle_spec[rel_dim] == Cycle::FIXED)
```
line length

devreal reviewed (changes requested):
Good work! A bunch of comments inline. 

In general: the code is basically undocumented, esp. the parts dealing with dimensions, offsets, and indexes could use some documentation, e.g., `HaloBlock::HaloBlock` among others.

dhinf committed:
resolved wrong indentions

dhinf committed:
- HaloSpec: replaced vector with array
- fixed wrong indentions in Halo.h and HaloMatrixWrapper.h
- RegionIter: removed _size
              removed operator + and - for type iterator
- HaloBlock: added constexpr to NumDimensions
             added reference to region_t in method halo_size

dhinf committed:
- code formating
- meaningful name for index calculation constant 3
- minor stuff

devreal commented on dash/include/dash/halo/Halo.h:
```diff
@@ -0,0 +1,1104 @@
+  static_assert( std::is_integral<ExpT>::value && std::is_unsigned<ExpT>::value,
+                 "Exponent must be an unsigned integer." );
+
+  return ( exp == 0 ? 1 : base * pow( base, exp - 1 ) );
```
`return ( exp == 0 ? 1 : (exp > 0) ? base * pow( base, exp - 1 ) : base / pow(base, exp + 1) );` would get rid of the requirement that `ExpT` be an unsigned int, which it wouldn't be if I called `pow(2, 5)`.

dhinf replied:
base / pow(base, exp + 1) ? -> that's not 1/ base^exp

devreal replied:
Gnarf, you're right! How about that: 

```
return ( exp == 0 ? 1 : (exp > 0) ? base * pow( base, exp - 1 ) : 1 / (base*pow(base, -(exp + 1))) );
```

dhinf replied:
```c++
return ( exp == 0 ? 1 : (exp > 0) ? base * pow( base, exp - 1 ) : pow(base, (exp + 1)) / base );
```
also works
but the result is the same: 0 -> return type is an integer

devreal replied:
Correct, so you can actually short cut it. This should be caught when moving it to `dash::ce`

devreal commented on dash/include/dash/halo/Halo.h:
```diff
@@ -0,0 +1,1104 @@
+   *
+   * \see DashGlobalIteratorConcept
+   */
+  reference operator[]( index_type idx ) const {
```
I'm a bit puzzled by the semantics of `operator[]` on an iterator: My understanding is that the iterator points to an element in a container. This implementation ignores the current position completely. If it's not really needed, I'd suggest to remove it.

dhinf replied:
We are using the subscript operator in other iterators the same way. The standard library uses this operator for the random access iterator different -> http://en.cppreference.com/w/cpp/concept/RandomAccessIterator

devreal replied:
`*(i + n)` seems to be reasonable if you think in pointer arithmetics. If we use the subscript operator in other places the same way those places need fixing as well. That's all but intuitive!

devreal commented on dash/include/dash/halo/Halo.h:
```diff
@@ -0,0 +1,1104 @@
+    std::array<std::pair<halo_extent_t, halo_extent_t>, NumDimensions>;
+  using size_type     = typename PatternT::size_type;
+  using CoordsT       = std::array<index_type, NumDimensions>;
+  using RegionCoordsT = RegionCoords<NumDimensions>;
```
Type naming is inconsistent: `SelfT` vs `size_type` vs `region_t`. I know we're sloppy throughout DASH with this but let's try to be consistent within one class at least  (helps readability) :)

devreal commented on dash/include/dash/halo/Halo.h:
```diff
@@ -0,0 +1,1104 @@
+    return _boundary_elements;
+  }
+
+  size_type halo_size() const {
```
it seems that the size can be computed once during setup and cached.

dhinf replied:
true

devreal commented on dash/include/dash/halo/Halo.h:
```diff
@@ -0,0 +1,1104 @@
+    }
+  }
+
+  ElementT* haloPos( region_index_t index ) { return _halo_offsets[index]; }
```
Also: you're using both underscores and CamelCase in the public interface, I think we should consolidate that to underscores.

devreal reviewed (changes requested):
Thanks for the fixes. I have some more comments now that I looked over it again. 

Are you planning to add documentation for this PR?

dhinf committed:
changed method names from ChamelCase to underscore devided

dhinf committed:
- consistent naming for template parameters and aliases

template:
  - classes -> ClassNameT
  - typed paramter TypedParamter
alias:
  - classes -> ClassName_t
  - simple type -> simple_type_t

exceptions are:
  - iterator traits -> standard library
  - own iterator and pointer non const/const

dhinf committed:
changed subscript operators for RegionIter and HaloMatrixWrapper to standard library behavior -> *(idx + n)

dhinf committed:
pow for integer moved from Halo.h to utils/FunctionalExpr.h

devreal commented on dash/include/dash/halo/Halo.h:
```diff
@@ -281,137 +277,141 @@ std::ostream& operator<<( std::ostream&                        os,
+  using Stencil_t       = Stencil<NumDimensions>;
+  using region_index_t  = typename RegionCoords_t::region_index_t;
+  using region_size_t   = typename RegionCoords_t::region_index_t;
+  using region_extent_t = typename RegionSpec_t::region_extent_t;
```
The types are declared private but exposed to the user through member functions. In that case, they should be public (I know this is messy as well all around DASH). Please also check other places for this.

devreal commented on dash/include/dash/halo/Halo.h:
```diff
@@ -281,137 +277,141 @@ std::ostream& operator<<( std::ostream&                        os,
     GlobStaticMem<ElementT, dash::allocator::SymmetricAllocator<ElementT>>;
-  using ViewSpecT = typename PatternT::viewspec_type;
+  using ViewSpec_t      = typename PatternT::viewspec_type;
+  using pattern_index_t = typename PatternT::index_type;
```
why `pattern_index_t` and not just `index_t`? Is there another (conflicting) index type?

dhinf replied:
The thing is, that if i use index_t, i also have to use size_t. Then it might be not clear whether it is std::size_t or an other. it's an ugly name clash. i also use region_size_t.

devreal replied:
Gnarf, how about `size_type` and `index_type`?

dhinf replied:
Personally i hate type as ending. Typically _t is used to indicate a type. Especially longer names don't profit from type.

devreal commented on dash/include/dash/halo/Halo.h:
```diff
@@ -898,205 +901,203 @@ class HaloBlock {
+  const ViewSpec_t& view() const { return _view; }
 
-  const ViewSpecT& view_safe() const { return _view_safe; }
+  const ViewSpec_t& view_safe() const { return _view_safe; }
```
Just out of curiosity: you mentioned that you are looking for a different name for `safe`, any luck so far?

dhinf replied:
view_foolproof? :)

dhinf replied:
but serious: how view_reliable sound?

devreal replied:
Right now, I can't think of the exact meaning of the safe view anymore (missing documentation, y'a know ;)). I vaguely remember that this was the view on the halo cells (?) Why not call it that?

dhinf replied:
is inner + boundary to guarantee neighbours -> thats why it's called view_guaranteed now

devreal reviewed (changes requested):
Looks good, just one more thing regarding the types :)

dhinf committed:
moved used aliases in methods to public and view_save is named view_guaranteed now

dhinf committed:
small fixes

dhinf committed:
view_guaranteed renamed to view_inner_with_boundaries

dhinf committed:
changed function call to view_local

devreal committed:
Merge branch 'development' into feat-halo
